README.md
# WCS | Windows Control Server

WCS, or Windows Control Server, is a tool that allows you to perform certain tasks on your Windows PC remotely from another device. So far the tool offers only features related to shutting or turning the machine off in a number of ways, accessing and modifying the file system and launching certain applications.

The tool itself represents a server that runs on the machine that you want to be able to perform certain tasks on, and a web client that you can access from any device on your LAN or worldwide if you configure your router's NAT with your ISP.

The web client is a regular frontend, with Sass code which is compiled into native CSS code on the server side (the Sass code comes precompiled with the project). The web client is also protected with a 4 to 8 digit PIN-code that you configure when setting up the server for the first time. The PIN-code is required on every request made to the server in order to access the control panel from where you can perform the tasks. The PIN-code is stored on the server encrypted using the AES 256 bit encryption algorithm.

The web client control panel has a set of buttons each designated to perform a specific task. Upon clicking on a button, an AJAX request is being made to the server. Based on what task you chose to perform a request will be made to the task-specific-URL.

Upon receiving a request to the task-specific-URL, the server will execute a batch file containing the relevant commands for the task.

All batch files can be found in the `server\batch` directory, each file given the semantically relevant name.

The server is written in Node.js, meaning the machine the server will be run on must have Node.js installed. There are also specific modules that are required for running the server, but those can be installed by running the `setup.cmd` file which is located in the root directory of the project.

<table>
	<tr>
		<th>
			<span>Table of Contents</span>
		</th>
	</tr>
	<tr>
		<td>
			<ol>
				<li><a href="#installing-git">Installing Git</a></li>
				<li><a href="#installing-nodejs">Installing Node.js</a></li>
				<li><a href="#installing-required-modules">Installing required modules</a></li>
				<li><a href="#cloning-the-repository">Cloning the repository</a></li>
				<li><a href="#initiating-the-server">Initiating the server</a></li>
				<li><a href="#configuring-the-server">Configuring the server</a></li>
				<ol>
					<li><a href="#changing-the-port">Changing the port</a></li>
					<li><a href="#setting-a-new-pin-code-and-port">Setting a new PIN-code (and port)</a></li>
				</ol>
				<li><a href="#adding-new-tasks">Adding new tasks</a></li>
				<li><a href="#compiling-the-sass-code">Compiling the Sass code</a></li>
			</ol>
		</td>
	</tr>
</table>

## Installing Git

You will be needing Git to clone this repository to your machine. You can download and install Git from their official website [here](https://git-scm.com/download/win).

## Installing Node.js

As mentioned, the server is written in Node.js, requiring its presence on the machine. You can download and install Node.js from their official website [here](https://nodejs.org/en/download/).

## Installing required modules

Some modules do not come preinstalled with Node.js (of which I noticed are the `ip` and `uuid` modules). You can install all modules that the server uses by running the `setup.cmd` file in the root directory of the project. The `setup.cmd` file will also install `nodemon`, a handy tool that automates the process of manually restarting the server upon file changes. It will also install `sass`, the style sheets that WCS uses. You technically don't need Sass on your machine if you don't plan on modifying the `server\client\sass\stylesheet.scss` file to you preferences. Otherwise you can find instructions on compiling the SCSS code into native CSS code at the end of this readme file.

## Cloning the repository

To clone the repository to your machine, go to the directory you wish to clone it to in your terminal, and run the following command:

```bat
git clone https://github.com/msfninja/wcs.git
```

This will clone the WCS repository to your machine. You can access the WCS directory from your terminal or explorer.

## Initiating the server

The server CLI is initiated by going to the `server` directory of the cloned repository in your terminal, and running the following command:

```bat
nodemon server.js
```

This will start the server CLI, but not make the HTTP server itself online. That will happen after a few more steps.

If everything went correctly, you should see this text:

```plain
	Welcome to WCS. This appears to be your first time using WCS.
	Provide a PIN-code to protect the control panel (4 to 8 digits).
	
	PIN-code:
```

Enter a PIN-code that's 4 to 8 digits, and hit Enter. The PIN-code will be immedtiately encrypted using the AES 256 bit encryption algorithm and the PIN-code hash will be written to the `server\root\dat\key.hash` file.

After providing a valid PIN-code and hitting Enter, you should see this text:

```plain
	Enter a port WCS will run on (a number from 0 to 64738, use 80 for no port).
	
	Port:
```

Enter a valid port, and hit Enter. The port you entered will be written to the `config.json` file and used in the future. Wish you change the port, simply change the value in the `config.json` file which is located in the root directory of the project, save the changes and restart the server (if it doesn't restart automatically, e.g. if you initiated the `server.js` file with the `nodemon` command with the argument/flag `--ignore` set to `'*.json'`).

After providing a valid port and hitting Enter, you should see this text (where &lt;your_ip&gt; is your machine's IP address, i.e., where the server will be accessible from on other devices, and &lt;your_port&gt; being the port that you provided in the previous step):

```plain
	WCS is running at the following address:
	http://<your_ip>:<your_port>
	
	Press ^C any time to shut the server down.
```

To see if the server actually works, copy the address from the terminal and paste in your browser to visit it. If everything went fine, you should see this screen in your browser:

<img src="https://github.com/msfninja/wcs/raw/main/screenshots/Landing%20page%20when%20accessing%20the%20WCS%20web%20client.png" alt="Landing page when accessing the WCS web client" width="80%" />

Enter the PIN-code that you provided in the terminal, and log in. You should see this screen now:

<img src="https://github.com/msfninja/wcs/raw/main/screenshots/Landing%20page%20after%20logging%20in%20with%20your%20PIN-code%20in%20the%20WCS%20web%20client.png" alt="Landing page after logging in with your PIN-code in the WCS web client" width="80%" />

You can press any button to perform a task and see if it works.

## Configuring the server

A few basic things of the server can be configured in the `config.json` file in the root directory of the project, such as the port the server runs on, the name of the application, and whether to ask for a new PIN-code and port every time you initiate the server CLI.

Things will definitely change in future releases, but as of now, configuration further than that requires modifying the source code of the server.

### Changing the port

The `config.json` file has a property called `port`, by default on line 7:

```json
"port": null,
```

You can change its value to any valid port any time and save the file for changes to apply (maybe you have to restart the server as well).

### Setting a new PIN-code (and port)

This can be achieved by simply making the server think that you are using it for the first time, and therefore prompting for a PIN-code, but that'll also prompt for a port afterwards. To make this happen, find the `sec` property, by default on line 8. This property contains another property called `init`:

```json
"sec": {
    "init": false
}
```

If its value is set to `true`, it won't ask for a PIN-code and a port upon initiating the server CLI. If it's set to `false` however, it will. Keep in mind that the server will automatically set this value to `true` every time you initiate the server CLI, enter a PIN-code and a port.

## Adding new tasks

There will be more tasks added upon each new release of WCS, but you can add your own too. The only way to add your own task is to simply create a batch file with your script. You can put the batch file in a new directory inside the `server\batch` directory to make it appear as a separate category in the control panel. There are a few guidelines/rules on how to name a directory:

 - Can consist of an unlimited amount of characters
 - Can start with and contain any alphanumerical character
 - Cannot contain non-alphanumerical characters
 - Use a dash (`-`) for spaces (whitespace characters will be ignored)
 - Text case will be ignored, the server will make every first character of a word (words are separated with a dash) upper case and all other characters lower case

Create a new batch file inside the `server\batch\<your_directory>` directory, write the script you want to be executed, and save the file. A few guidelines/rules on how to name a batch file:

 - Can consist of an unlimited amount of characters
 - Can start with and contain any alphanumerical character
 - Cannot contain non-alphanumerical characters
 - Text case will be ignored, the server will make the first character upper case, and the rest lower case).

Depending on how you name the file, the name will appear in the control panel.

A sample batch script:

```bat
echo Hello, World!
```

Saved batch file in explorer:

<img src="https://github.com/msfninja/wcs/raw/main/screenshots/A%20batch%20file%20in%20the%20batch%20directory%20in%20WCS%20in%20explorer.png" alt="A batch file in the batch directory in WCS in explorer" width="80%" />

That's it.

The button will appear in the web client under the category what you named the directory the batch file is in to in the `server\batch` directory.

## Compiling the Sass code

WCS uses Sass. The actual file itself is written in SCSS though, not in SASS, due to syntax preferences. The SCSS file is located in the `server\client\sass` directory. If you go to the project's root directory in your terminal, you can run this command to compile the SCSS code once:

```bat
sass .\server\client\sass\stylesheet.scss .\public\client\css\stylesheet.css
```

Or use run this command to make Sass compile the SCSS code every time you make changes to it:

```bat
sass -w .\server\client\sass\stylesheet.scss .\public\client\css\stylesheet.css
```

---

&#169; 2021 WCS, Windows Control Server

CHANGELOG
Release v0.0.0 2021-09-22 msfninja <msfninja@airmail.cc>

	Note:
	Several general fixes are introduced, mainly regarding the documentation and the README.md file. The server.js file's source code had been partly minified and small improvements have been introduced as well. The web client has also met new features and improvements.

	Added:
	 * New features/tasks were added to the web client
	   - Ability to launch a custom batch script
	   - Ability to launch applications
	   - Ability to modify the file system
	   - Ability to retrieve files from the file system

	Changed:
	 * The /server/server.js file has been modified
	 * The /server/root/index.html file has been modified
	 * The /README.md file has been modified
	 * The /setup.bat file has been renamed to /setup.cmd
	 * the /setup.cmd file has been modified
	 * The "A batch file in the batch directory in WCS in explorer.png" screenshot has been cropped 1 pixel from both horizontal sides
	 * The "Landing page when accessing the WCS web client.png" and "Landing page after logging in with your PIN-code in the WCS web client.png" screenshota have been updated

	Fixed:
	 * Function names and line numbers were corrected in the README.md file

Pre-release v0.0.0-alpha 2021-09-14 msfninja <msfninja@airmail.cc>
	
	Note:
	Project initiated.

server/server.js
const
	http = require('http'),
	path = require('path'),
	url = require('url'),
	fs = require('fs'),
	qs = require('querystring'),
	ip = require('ip'),
	colors = require('colors'),
	{ exec } = require('child_process'),
	crypto = require('crypto'),
	{
		v1: uuidv1,
		v4: uuidv4
	} = require('uuid'),
	readline = require('readline'),
	rl = readline.createInterface({
		input: process.stdin,
		output: process.stdout
	});

const
	algorithm = 'aes-256-ctr',
	iv = crypto.randomBytes(16);

const dir = `${__dirname}/..`;

const
	ex = h => { return fs.existsSync(h); },
	rd = h => { return ex(h) ? fs.readFileSync(h).toString() : ''; },
	wr = (h,c) => { fs.writeFile(h,c,(err) => { if (err) throw err; }); };

const
	log = t => {
		console.log(`\n\t${t.replace(/\n/g,'\n\t')}`);
	},
	clear = () => { console.clear(); },
	error = t => {
		console.error(`\n\tERROR!\n\n\t${t.replace(/\n/g,'\n\t')}`);
		process.exit(0);
	};

const config = JSON.parse(rd(`${dir}/config.json`));

const cli = () => {
	let root = new Root(); clear();
	if (!root.exists()) {
		log(`Welcome to ${config.app.name}. This appears to be your first time using ${config.app.name}.\nProvide a PIN-code to protect the control panel (4 to 8 digits).\n`);
		rl.question('\tPIN-code: ',pin => {
			if (!isNaN(pin) && pin.split('').length >= 4 && pin.split('').length <= 8) {
				root.newcred(pin);
				clear();
				log(`Enter a port ${config.app.name} will run on (a number from 0 to 64738, use 80 for no port).\n`);
				rl.question('\tPort: ',port => {
					if (!isNaN(port) && parseInt(port) >= 0 && parseInt(port) <= 64738) {
						config.port = port;
						config.sec.init = true;
						wr(`${dir}/config.json`,JSON.stringify(config,null,4));
						start(port);
						rl.close();
					}
					else {
						log('Invalid port.');
						rl.close();
						process.exit();
					}
				});
			}
			else {
				log('Invalid PIN.');
				rl.close();
				process.exit();
			}
		});
	}
	else start(config.port);
};

const
	t = (res,s,h,c) => {
		res.writeHead(s,h);
		if (c) res.write(c);
		res.end();
	},
	render = (res,h,m) => {
		var
			modes = m ? m.split('') : [],
			html = '';
		if (compare(['r','a'],modes,1)) {
			html = rd(h).replace(/\{component\.meta\}/g,rd(`${dir}/server/client/component/meta.html`)).replace(/\{component\.links\}/g,rd(`${dir}/server/client/component/links.html`)).replace(/\{component\.scripts\}/g,rd(`${dir}/server/client/component/scripts.html`));
			if (modes.includes('a')) {
				html = html.replace(/\{categories\}/g,gethtml());
				html = html.replace(/\{token\}/g,token);
			}
			html = html.replace(/\{app\.name\}/g,config.app.name).replace(/\{app\.namelong\}/g,config.app.name_long);
		}
		return html;
	},
	compare = (a,b,n) => {
		var arr = [];
		for (var i = a.length - 1; i >= 0; i--) for (var j = b.length - 1; j >= 0; j--) if (a[i] === b[j]) arr.push(a);
		if (arr.length >= n) return true;
		return false;
	},
	cookies = r => {
		var
			obj = {},
			rc = r.headers.cookie;
		rc && rc.split(';').forEach(e => {
			var parts = e.split('=');
			obj[parts.shift().trim()] = decodeURIComponent(parts.join('='));
		});
		return obj;
	},
	check = () => {
		var arr = [
			'/server/root/dat/uuid.asc',
			'/server/root/dat/key.hash',
			'/public/index.html',
			'/server/root/index.html'
		];
		arr.forEach(e => { if (!ex(`${dir}${e}`)) return false; });
		return true;
	};

const
	gettree = h => { return fs.readdirSync(h); },
	getcom = p => {
		var arr = [];
		const getlist = p => {
			gettree(p).forEach(e => {
				if (fs.lstatSync(path.join(p,e)).isDirectory()) getlist(path.join(p,e));
				else arr.push({
					name: e.split('.')[0],
					path: path.join(p,e)
				});
			});
		};
		getlist(p);
		return arr;
	},
	getcat = p => {
		var arr = [];
		gettree(p).forEach(e => { if (fs.lstatSync(path.join(p,e)).isDirectory()) arr.push(e); });
		return arr;
	},
	execute = (a,b,c) => {
		var obj = getcom(`${dir}/server/batch`).find(e => e.name === a);
		if (obj) {
			var str = obj.path
			if (b) str += ` ${decodeURIComponent(b)}`;
			else if (c) str += ` ${decodeURIComponent(c)}`;
			exec(str);
		}
	};

const gethtml = () => {
	var html = '';
	getcat(`${dir}/server/batch`).forEach(cat => {
		html += `
			<section>
				<div class="h4">${formatcat(cat)}</div>
				<div class="ws50"></div>
				<div class="scrollbox-h">
		`;
		getcom(`${dir}/server/batch`).forEach(com => {
			if (cat === 'file-system' && com.path.split(path.sep)[com.path.split(path.sep).length - 2] === 'file-system') {
				html += `
					<script type="application/ecmascript" defer="defer">
						const fs${com.name} = () => {
							var c = prompt('${com.name}: Enter a pathname');
							if (c) {
								ajax(\`/{token}/${com.name}/\${encodeURIComponent(c)}\`);
								notif(true,'Command','${com.name} command sent');
							}
						};
					</script>
					<div title="Send ${com.name} command to your PC" onclick="fs${com.name}();" class="btn">
						<div>${com.name}</div>
					</div>
				`;
			}
			else if (com.path.split(path.sep)[com.path.split(path.sep).length - 2] === cat) {
				html += `
					<div title="Send ${com.name} command to your PC" onclick="ajax('/{token}/${com.name}'); notif(true,'Command','${formatcom(com.name)} command sent');" class="btn">
						<div>${formatcom(com.name)}</div>
					</div>
				`;
			}
		});
		html += `
				</div>
			</section>
		`;
	});
	return html;
};

const
	formatcat = str => {
		var arr = [];
		str.split('-').forEach(e => { arr.push(e.charAt(0).toUpperCase() + e.slice(1).toLowerCase()); });
		return arr.join(' ');
	},
	formatcom = str => { return str.charAt(0).toUpperCase() + str.slice(1).toLowerCase(); };

function Root(res) {
	this.access = c => {
		if (c) t(res,200,{'Content-Type':'text/html'},render(res,`${dir}/server/root/index.html`,'ra'));
		else t(res,200,{'Content-Type':'text/html'},render(res,`${dir}/public/index.html`,'r'));
	}
	this.exists = () => {
		return (config.sec.init && rd(`${dir}/server/root/dat/uuid.asc`).split('').length === 32);
	};
	this.verify = p => {
		if (cred() && pinuuid()) {
			if (p === decrypt(cred().pin,(`${p}${pinuuid()}`).slice(0,32))) return true;
			return false;
		}
	};
	this.newcred = p => {
		if (pinuuid()) wr(`${dir}/server/root/dat/key.hash`,JSON.stringify({ pin: encrypt(p,(`${p}${pinuuid()}`).slice(0,32)) }));
		else error(`The uuid.asc file seems to be misconfigured or missing. Cannot create new credentials.\npinuuid() response:\n\n${pinuuid()}`);
	};
}

const start = port => {
	http.createServer((req,res) => {
		var
			q = url.parse(req.url,true),
			p = q.pathname,
			cookie = cookies(req);
		let
			root = new Root(res);
		if (['/'].includes(p)) {
			if (req.method === 'POST') {
				let body;
				req.on('data',chunk => body = chunk.toString());
				req.on('end',() => {
					let post = qs.parse(body);
					if (post.pin && root.verify(post.pin)) root.access(true);
					else root.access(false);
				});
			}
			else t(res,200,{'Content-Type':'text/html'},render(res,`${dir}/public/index.html`,'r'));
		}
		else if ([token].includes(p.split('/')[1])) execute(p.split('/')[2],p.split('/')[3]);
		else {
			const h = path.resolve(dir,p.replace(/^\/*/,`${dir}/public/`));
			if (ex(h)) {
				fs.readFile(h,(err,dat) => {
					if (err) t(res,404,{'Content-Type':'text/html'},render(res,`${__dirname}/client/err/404.html`,'r'));
					res.statusCode = 200;
					return res.end(dat);
				});
			}
			else t(res,404,{'Content-Type':'text/html'},render(res,`${__dirname}/client/err/404.html`,'r'));
		}
	}).listen(port);
	clear();
	log(`${config.app.name} is running at the following address:\nhttp://${ip.address()}${(port == 80) ? '' : (':' + port)}\n\nPress ^C any time to shut the server down.`);
};

const
	encrypt = (t,k) => {
		const
			cipher = crypto.createCipheriv(algorithm,k,iv),
			encrypted = Buffer.concat([cipher.update(t),cipher.final()]);
		return {
			iv: iv.toString('hex'),
			content: encrypted.toString('hex')
		};
	},
	decrypt = (h,k) => {
		const
			decipher = crypto.createDecipheriv(algorithm,k,Buffer.from(h.iv,'hex')),
			decrpyted = Buffer.concat([decipher.update(Buffer.from(h.content,'hex')),decipher.final()]);
		return decrpyted.toString();
	};

const
	cred = () => {
		try { return JSON.parse(rd(`${dir}/server/root/dat/key.hash`)); }
		catch (err) { return false; }
	},
	pinuuid = () => {
		try {
			if (rd(`${dir}/server/root/dat/uuid.asc`).split('').length == 32) {
				return rd(`${dir}/server/root/dat/uuid.asc`);
			}
			else {
				let key = getuuid(1);
				wr(`${dir}/server/root/dat/uuid.asc`,key);
				return key;
			}
		}
		catch (err) { return false; }
	};

const getuuid = n => {
	var str = '';
	for (var i = 0; i < n; i++) str += uuidv4().toString().replace(/\-/g,'');
	return str;
};

let token = getuuid(3);

if(check()) {
	try { cli(); }
	catch (err) { error(`Failed to initiate the commandline interface with the following error:\n\n${err}`); }
}
else error(`It appears that crucial files are missing. Cannot initiate server.\n\nTry reinstalling ${config.app.name} or contacting the creator for help.`);